src/udp_mux/mod.rs

```diff
@@ -0,0 +1,332 @@
+    async fn create_muxed_conn(self: &Arc<Self>, ufrag: &str) -> UDPMuxConn {
+        let local_addr = self.params.udp_socket.local_addr().ok();
+
+        let params = UDPMuxConnParams {
+            local_addr,
+            key: ufrag.into(),
+            udp_mux: Arc::clone(self),
+        };
+
+        UDPMuxConn::new(params)
+    }
+
+    fn register_conn_for_address(&self, conn: &UDPMuxConn, addr: SocketAddr) {
+        if self.is_closed() {
+            return;
+        }
+
+        let key = conn.key();
+        {
+            let mut addresses = self
+                .address_map
+                .write()
+                .expect("Failed to obtain write lock");
+
+            addresses
+                .entry(addr)
+                .and_modify(|e| {
+                    e.remove_address(&addr);
+                    *e = conn.clone()
+                })
+                .or_insert_with(|| conn.clone());
+        }
+
+        log::debug!("Registered {} for {}", addr, key);
+    }
+
+    async fn conn_from_stun_message(&self, buffer: &[u8], addr: &SocketAddr) -> Option<UDPMuxConn> {
+        let (result, message) = {
+            let mut m = STUNMessage::new();
+
+            (m.unmarshal_binary(buffer), m)
+        };
+
+        match result {
+            Err(err) => {
+                log::warn!("Failed to handle decode ICE from {}: {}", addr, err);
+                None
+            }
+            Ok(_) => {
+                let (attr, found) = message.attributes.get(ATTR_USERNAME);
+                if !found {
+                    log::warn!("No username attribute in STUN message from {}", &addr);
+                    return None;
+                }
+
+                let s = match String::from_utf8(attr.value) {
+                    // Per the RFC this shouldn't happen
+                    // https://datatracker.ietf.org/doc/html/rfc5389#section-15.3
+                    Err(err) => {
+                        log::warn!(
+                            "Failed to decode USERNAME from STUN message as UTF-8: {}",
+                            err
+                        );
+                        return None;
+                    }
+                    Ok(s) => s,
+                };
+
+                let conns = self.conns.lock().await;
+                let conn = s
+                    .split(':')
+                    .next()
+                    .and_then(|ufrag| conns.get(ufrag))
+                    .map(Clone::clone);
+
+                conn
+            }
+        }
+    }
+
+    fn start_conn_worker(self: Arc<Self>) {
+        tokio::spawn(async move {
+            let mut buffer = [0u8; RECEIVE_MTU];
+
+            loop {
+                let loop_self = Arc::clone(&self);
+                let socket = &loop_self.params.udp_socket;
+
+                if loop_self.is_closed() {
+                    break;
+                }
+
+                match socket.recv_from(&mut buffer).await {
```

Comment:
let's use closed channel, instead of AtomicBool, so that we can use
```
tokio::select!{
    _ = closed_rx.recv() => {
          break;
    }
    res = socket.recv_from....
}
```
otherwise, it will be blocked in socket.recv_from and no chance to check closed